### PR TITLE
Kick reason

### PR DESCRIPTION
**Pull request by wolfreak99**
I have now finished editing this base and am adding some final touches to the code, then will likely squash all commits together. The idea of this branch / pull request is to provide a dropdown that extends the reason you are kicked from a server. You are given a few select predefined reasons, along with the option to enter your own reason. Instead of a player seeing "disconnected: get out of the server!" they will see "disconnected: vandalism (wolfreak_99)" (in the event that i was to kick them, for example. The idea of this is to provide further reason as to why someone would be kicked if preferred. The predefined list also has "No reason" selected by default that way if you click the kick button, it will behave like normal: it will say the generic "get out of the server". The provided reasons are "No reason, vandalism, spam, harassment, illegal content, Other reason", The other reason will bring up the text input box allowing the person to type in a response.

please note that if you are testing this, you will only be able to do this in your own server or a server that has this kick reason implemented, all outside kick attempts will fail because this replaces the built in game command version.

### CODE REVIEW

**Comment by Nubbie** [comment]
My suggestion for pre-made messages would be:

- Greifing
- Harassment
- Spam
- Editing coasters/stores
- Killing Guests
- Breaking this servers rules
- Inappropriate behavior
- Other

With some really clever pre-made strings, it wouldn't be necessary with a custom message, especially when I'm really fairly sure they will be used in more negative ways than positive
 I did see this in action on your server for example which made me think otherwise, really don't think custom messages will be used for good. :worried:

**Comment by wolfreak99** [comment]
"I did see this in action on your server for example which made me think otherwise" Oh yeah... not even gonna lie, i abused it... I loved doing it too.. lol.. 

Do you think griefing would be a better word for vandalism? like, personally i feel vandalism gets the same idea as griefing, but if people think griefing would be better suited for this, i will be more than happy to switch it.

if i do "Breaking this servers rules" i will likely shorten it to "breaking server rules", i like the idea of smaller messages simply because that box that opens up looks better when it's thin and not as wide with long text. 


one thing i want to address.. this idea has the name of who kicks you at the end. personally, i think it would help if someone wanted to come in and clear up what happened with whomever kicked them (should it be a misunderstanding or mistake), or it would help point out if someone goes on a kicking spree (and thus, if someone can revoke that persons privilages, they can do so with full assurance they're revoking the right person).. but i can also see people going "fuck you wolfreak why did you kick me" or something, or a negative outlash of sorts. do you think i should make this something the server host can see, or make it in a permission such as "see kicked by information" or something (i'll need help coming up with a name for the permission if it comes down to it), or if the general community thinks it wouldn't be that big of a deal and would be okay to keep? Should i also make it to where the server host can disable appending who kicked the user?

As far as removing the "other" option, if the majority is against it, i will remove it. otherwise, i will likely allow a server host setting that can enable custom messages (that will likely be disabled by default). i wanna wait to make sure the majority is against it and no one is for it though, because personally i loved being able to kick people for detailed reasons such as "did you really just edit my ride after i told you not to?" or something such as that, so removing it feels kinda hard haha.

**Comment by Gymnasiast** [comment]
There should be an 'Other' field. There are plenty of options for evil moderators to say nasty things to people; getting rid of the 'Other' reason won't do much while it _will_ affect many legitimate uses - if I ran a server, I'd want to tell people why they got banned exactly, rather than just slapping a generic reason on the ban. I would even say that the custom reason should always be available, to allow the moderator to give more details.

**Comment by wolfreak99** [comment]
Thank you for helping back me up. I like your point about just slapping a generic reason on it. How do you feel about the name being shown? (Oh, and one thing, you say the word "ban", it's still technically a "kick". in fact the reason i originally had the generic reasons was for the event that someone was to hop back in and continue vandalising, it'd be easy to kick them out again quickly, without having to type it out again... But soon you may be able to say the word ban as i'm hoping to get started on another branch for banning, and then integrate this, should it get accepted oncei'm fininshed with this)

File: src/network/NetworkAction.cpp [hunk 1]
```diff
@@ -191,7 +191,7 @@ const std::vector<NetworkAction> NetworkActions::Actions =
     }, {
         STR_ACTION_KICK_PLAYER, "PERMISSION_KICK_PLAYER",
         {
-            GAME_COMMAND_KICK_PLAYER
+            -4
```
Comment (Broxzier) [inline src/network/NetworkAction.cpp]: Where does this magic number come from?

Comment (wolfreak99) [inline src/network/NetworkAction.cpp]: @Broxzier The majority of the network commands have been negative thus far, i was pretty much just going with the flow

Comment (Broxzier) [inline src/network/NetworkAction.cpp]: @wolfreak100 I mean the number itself. But I just checked and saw there were `-1`, `-2`, and `-3` further in the list as well. It just seems odd to me someone would replace a macro with a magic number.

File: src/windows/player.c [hunk 2]
```diff
@@ -264,6 +265,65 @@ void window_player_overview_show_group_dropdown(rct_window *w, rct_widget *widge
 	dropdown_set_checked(network_get_group_index(network_get_player_group(player)), true);
 }
 
+#define KICK_REASON_INDEX_FIRST STR_KICK_REASON_NO_REASON
+#define KICK_REASON_INDEX_LAST STR_KICK_REASON_OTHER_REASON
+#define KICK_REASON_INDEX_COUNT (KICK_REASON_INDEX_LAST + 1 - KICK_REASON_INDEX_FIRST)
+#define KICK2STRING(x) (KICK_REASON_INDEX_FIRST + x)
+#define STRING2KICK(x) (x - KICK_REASON_INDEX_FIRST)
```
Comment (IntelOrca) [inline src/windows/player.c]: Its better to instead create a const array of string IDs. This is done in several places, like the keyboard shortcuts.

Comment (wolfreak99) [inline src/windows/player.c]: @IntelOrca "Its better to instead create a const array of string IDs. This is done in several places, like the keyboard shortcuts."

I can try to do this, though the idea of this way was an attempt to allow people to easily add new custom string reasons by only needing to adjust the string_id file and en-gb or whatnot file and as long as it was within the boundary of no_reason and other_reason, it'd show up perfectly. I can try your suggestion as my way did and still does feel too hacky/clunky. i haven't yet looked up on google about this or skimmed the code that much, but would there be a way to have the const array have a "kick_reason_count" at the end that doesn't point to a higher rct_string_id number, but he actual number in the array? or will i have to create both an enum and string_id array? or will i have to manually insert the 6 in "const rct_string_id KickReasonStringIDs[6] = {" ?

Comment (IntelOrca) [inline src/windows/player.c]: Once strings are added after other_reason, you won't be able to just insert new ones inbetween them, or you would be limited by the space you left. We will move away from fixed string IDs eventually and assuming sequential string IDs is not a good idea.

You will need to create an enum and an array, see keyboard shortcuts as an example.

File: src/network/network.cpp [hunk 3]
```diff
@@ -593,20 +594,32 @@ bool Network::CheckSRAND(uint32 tick, uint32 srand0)
...
+				gGameCommandErrorText = STR_CANT_KICK_THE_HOST;
+				return false;
+			}
 			// Disconnect the client gracefully
-			(*it)->SetLastDisconnectReason(STR_MULTIPLAYER_KICKED);
-			char str_disconnect_msg[256];
-			format_string(str_disconnect_msg, STR_MULTIPLAYER_KICKED_REASON, NULL);
+			char str_disconnect_msg[NETWORK_DISCONNECT_REASON_BUFFER_SIZE];
+			const char* kickedText = language_get_string(STR_MULTIPLAYER_KICKED);
+			sprintf_s(str_disconnect_msg, NETWORK_DISCONNECT_REASON_BUFFER_SIZE, "%s: %s (%s)", kickedText, reason, kickedBy);
```
Comment (Goddesen) [inline src/network/network.cpp]: This will only compile on windows. `snprintf` is the non-windows equivalent.

Comment (wolfreak99) [inline src/network/network.cpp]: Thanks bro, much appreciated.

File: src/windows/player.c [hunk 4]
```diff
@@ -264,6 +278,63 @@ void window_player_overview_show_group_dropdown(rct_window *w, rct_widget *widge
...
+	if (strReasonIndex == STR_KICK_REASON_NO_REASON) {
+		format_string(msg, STR_MULTIPLAYER_KICKED_REASON, NULL);
+	}
+	else if (strReasonIndex == STR_KICK_REASON_OTHER_REASON) {
+		if (!reason || strcmp(reason, "") == 0) {
+			format_string(msg, STR_MULTIPLAYER_KICKED_REASON, NULL);
+		}
+		else {
+#ifndef DISABLE_NETWORK
+			snprintf(msg, NETWORK_DISCONNECT_KICK_REASON_MAX_SIZE, reason);
```
Comment (Goddesen) [inline src/windows/player.c]: Never ever use any version of `printf` with a variable format argument. This is an easy attack vector for malicious purposes. In this case you can probably use `strncpy`, that is if you need the `msg` to be a separate pointer at all (I didn't look at the code very closely).

File: src/windows/player.c [hunk 5]
```diff
@@ -304,6 +375,19 @@ void window_player_overview_mouse_down(int widgetIndex, rct_window *w, rct_widge
...
+	}
+}
+
+void window_player_overview_textinput(rct_window *w, int widgetIndex, char* text)
+{
+	if (!text) {
+		return;
+	}
+	if (widgetIndex == WIDX_KICK) {
+		window_player_overview_kick_player(w, STR_KICK_REASON_OTHER_REASON, _strdup(text));
```
Comment (Goddesen) [inline src/windows/player.c]: This usage of `_strdup` will leak memory. I don't think you need a *copy* of `text` here?

File: src/windows/player.c (hunk 4 above)
Comment (wolfreak99) [inline src/windows/player.c]: i just need the value of msg to pass.the idea of this part is it chooses whether to use a predefined text (from the string id), or use a custom text, and then it sends that value over as a parameter) not sure whether it needs to be a pointer or not. i'm just trying to pass the string over.

File: src/windows/player.c (hunk 5 above)
Comment (wolfreak99) [inline src/windows/player.c]: how would i do that? do the same method as i did with the char msg[128] and strncpy?

File: src/windows/player.c (hunk 4 above)
Comment (Goddesen) [inline src/windows/player.c]: `snprintf` is for when you want to supply arguments that need to be formatted. `strncpy` should do fine.

File: src/windows/player.c (hunk 5 above)
Comment (Goddesen) [inline src/windows/player.c]: In this case you can just send `text` through. You are sending a pointer to the text, there is no need to copy the text anywhere in this function.

File: src/windows/player.c (hunk 4 above)
Comment (Goddesen) [inline src/windows/player.c]: Actually, I just noticed there is `safe_strcpy` in `util/util.c` that does this, but also a bunch of checks as well as handling utf8. The text input should probably be utf8 too... Could someone with knowledge on this point chime in?

**Comment by wolfreak99** [comment]
I think I need some help with this Travis crap. I just dont get what im doing wrong anymore.

**Comment by Gymnasiast** [comment]
The problem is that NETWORK_DISCONNECT_KICK_REASON_MAX_SIZE is not always declared - to be precise, it is not declared when you build without multiplayer support.

**Comment by wolfreak99** [comment]
Would I be able to move it above the ifndef network_disable line in network.h? I fear that might not really be wanted since the stream id and etc are within the network_disable and work just fine.

**Comment by wolfreak99** [comment]
What the fuck, now appveyor is failing and not even giving a reason why. I'm just going to close this and give up, cause this is becoming too much of a hassle for something I can even fully declare finished by heart.

**Pull request closed without merging**